server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -1834,6 +1846,71 @@ public long getDefaultKeepAliveInMillis() {
         return defaultKeepAlive;
     }
 
+    public List<ReaderContext> getActiveContexts(ShardId shardId) {
+        return this.activeReaders.values()
+            .stream()
+            .filter(c -> c.singleSession() == false)
+            .filter(c -> c.scrollContext() == null)
+            .filter(c -> c.indexShard().shardId().equals(shardId))
+            .collect(Collectors.toList());
+    }
+
+    public void reopenPitContexts(ShardId shardId, String segmentsFileName, long keepAlive, String sessionId, long contextId) {
```

Comment:
This whole part is pretty rough still, I mostly wanted _something_ working in the receiving side of the PIT relocation that makes the integration test work. There are probably many open questions here.

Reply:
I wonder if we could move part of this logic into the `Engine`? this would allow us to track the necessary metadata to keep blobs around (and it would throw if we ever try to call this in an `Engine` that doesn't support to transfer PIT contexts?